File: MMath work/4th Year/Uncertainty Quantification/Exercises/Week2Exercises.R
#### Ex.1 ####
Lklh<-function(x,n,mu){
  likelihood<-1
  for (i in 1:n){
    likelihood<-likelihood*(2*pi)^(-0.5)*exp(-0.5*(x[i]-mu)^2)
  }
  return(-likelihood)
}
n<-100
sampleHolder<-matrix(0,nrow=20,ncol=n)
for (j in 1:20){
  x<-rnorm(n,0.5,1)
  sampleHolder[j,]<-x
}
mu<-seq(-1,1,length.out=n)
Likelihood<-rep(0,length(mu))
LikelihoodHolder<-matrix(0,nrow=20,ncol=length(mu))
for (j in 1:20){
  for (i in 1:100){
    Likelihood[i]<-Lklh(sampleHolder[j,],n,mu[i])*(-1)
  }
  LikelihoodHolder[j,]<-Likelihood
  Likelihood<-rep(0,length(mu))
}
plot(mu,LikelihoodHolder[5,],col="blue",xlab="mu",ylab="L(mu,1)",cex=0.1)
lines(mu,LikelihoodHolder[1,])
lines(mu,LikelihoodHolder[2,])
lines(mu,LikelihoodHolder[3,])
lines(mu,LikelihoodHolder[4,])
lines(mu,LikelihoodHolder[5,])

mleHolder<-rep(0,5)
arithmeticMeanHolder<-rep(0,5)
for (j in 1:5){
  mle<-optim(par=c(0.5),Lklh,x=sampleHolder[j,],n=100)
  mleHolder[j]<-mle$par
  arithmeticMean<-sum(sampleHolder[j,])/n
  arithmeticMeanHolder[j]<-arithmeticMean
}
matplot(mu,t(LikelihoodHolder),type="l")

#### Ex.2 ####
logL<-function(x,n,mu,sigma){
  logLik<- (-1)*n*log(2*sigma)
  for (i in 1:n){
    logLik<-logLik-(1/(2*sigma))*abs(x[i]-mu)
  }
  return(logLik)
}
n<-50
x<-rLaplace(n,1.5,2)
mu<- -1
sigma<-seq(0.1,5,by=0.1)
Likelihood<-rep(0,length(sigma))
for (i in 1:length(sigma)){
  Likelihood[i]<-logL(x,n,mu,sigma[i])
}
plot(sigma,Likelihood,xlab="Sigma",ylab="log(Likelihood)",ylim=c(-250,0))
mu2<- seq(-1,1,by=0.2)
for (j in 1:length(mu2)){
  for (i in 1:length(sigma)){
    Likelihood[i]<-logL(x,n,mu2[j],sigma[i])
  }
  lines(sigma,Likelihood)
}
#seems that increasing mu tends to increase the mle value of sigma
#### Ex.3 ####
n<-seq(1,1000,by=20) #create vector of n values to test
Error<-rep(0,length(n)) #to hold the errors for each n
for (i in 1:length(n)){
  x<-rbinom(i,1,0.6) #generate random sample
  theta_m<-sum(x)/i #find arithmetic mean
  Error[i]<-abs(theta_m-0.6) #find absolute error
}
plot(n,Error,xlab="n",ylab="E(n)") #produce plot
nstar<-250
counter<-0
for (i in 1:10000){
  x<-rbinom(nstar,1,0.6)
  theta_m<-sum(x)/nstar
  Jtheta<-sum(x)/(theta_m^2)+(nstar-sum(x))/((1-theta_m)^2)
  lowCI<-theta_m-1.96/sqrt(Jtheta) #calculate lower bound
  upCI<-theta_m+1.96/sqrt(Jtheta) #calculate upper bound
  if (lowCI<0.6 && upCI>0.6){ #add 1 if 0.6 between bounds
    counter<-counter+1
  }
}
proportion<-counter/10000 #find proportion
proportion
#### Ex.4 ####
theta<-runif(1,0,1) #get random theta from uniform [0,1]
theta
n<-50
D<-rbinom(n,1,theta)
x<-seq(0,1,by=0.01)
BetaForPlot<-dbeta(x,1+sum(D),n-sum(D))
plot(BetaForPlot~x,col="blue",lty=1)
x[which(BetaForPlot==max(BetaForPlot))]
#Get a second interval (credible)
b<-qbeta(0.98,1+sum(D),n-sum(D)) #use in-built function at certain quantiles with difference 0.95
a<-qbeta(0.03,1+sum(D),n-sum(D))
FindInterval<-rbeta(10000,1+sum(D),n-sum(D)) #find interval from posterior distribution sample
FindInterval<-sort(FindInterval)
a<-FindInterval[10000*0.025]
b<-FindInterval[10000*0.975]
#### Ex.5 ####
n<-seq(20,60,by=20)
mu<-0
alpha<-4
beta<-5
supportx<-seq(0.1,2.1,by=0.01)
betas<-matrix(0,nrow=3,ncol=length(supportx))
for (i in n){
  x<-rnorm(i,mu,1) #tau=1
  sum<-0
  for (j in 1:i){
    sum<-sum+(x[j]-mu)^2
  }
  betanew<-beta+0.5*sum
  for (k in 1:length(supportx)){
    betas[i/20,k]<-dbeta(supportx[k],alpha,betanew)
  }
}
plot(supportx,betas[1,],col="blue",xlab="tau",ylab="Probability")
lines(supportx,betas[2,],col="red")
lines(supportx,betas[3,],col="orange")

File: MMath work/4th Year/Uncertainty Quantification/Exercises/Week2Exercises.m
x=normrnd(0.5,1,[1,500]);
mu=linspace(-1,1,100);
sum=zeros(1,length(mu));
logL=zeros(1,length(mu));
for i=1:length(mu)
    for k=1:length(x)
        sum(i)=sum(i)+(x(k)-mu(i))^2;
    end
    logL(i)=-10*log(2*pi)-0.5*sum(i);
end
plot(mu,logL,'-b');
xlabel('Mu');
ylabel('logL(mu,1)');
mle=mu(logL==max(logL));
n=length(x);
analyticMle=mean(x);
disp(['When n=',num2str(n),' the maximum of the log-likelihood is ',num2str(mle), ' and the analytical mle (mean) is ',num2str(analyticMle)]);

%% Exercise 3 Part 1
n=linspace(50,10000,996);
error=zeros(1,length(n));
for i=50:10:10000
    p=zeros(1,i);
    for j=1:i
        p(j)=binornd(1,0.6); %Theta specified here
    end
    thetam=mean(p);
    error(i/10-4)=abs(0.6-thetam);
end
plot(n,error,'-b');
xlabel('n');
ylabel('Error');
%% Exercise 3 part 2
nstar=500;
N=100000;
counter=0;
theta=0.6;
for i=1:N
   rng(i);
   p=binornd(1,theta,[1,nstar]);
   [~,pci]=mle(p,'distribution','Bernoulli');
   if theta>pci(1) && theta<pci(2)
        counter=counter+1;
    end
end
disp(counter);
proportion=counter/N;
disp(['The proportion of confidence intervals containing the true value of theta=0.6 is ',num2str(proportion)]);
%% Exercise 4
theta=unifrnd(0,1);
disp(theta);
simN=zeros(1,1000);
for i=1:1000
    simN(i)=binornd(1,theta);
end
k=1000*mean(simN);
beta1=1+k;
beta2=1000-k;
thetaArg=0:0.01:1;
pifunc=betapdf(thetaArg,beta1,beta2);
plot(thetaArg,pifunc,'b');
xlabel('Theta');
ylabel('Pi(Theta given D)');
p=[0.025,0.975];
x=betainv(p,beta1,beta2);
width=diff(x);
disp(['The interval with theta=',num2str(theta),' is']);
disp(x);
disp(['The interval has width ',num2str(width)]);
N=1000;
samp=betarnd(beta1,beta2,1,N);
samp=sort(samp);
a=samp(N*0.025);
b=samp(N*0.975);
MCinterval=[a b];
disp('Interval from Monte Carlo is');
disp(MCinterval);

%% Exercise 5
alpha=5;
beta=2;
tau1=gamrnd(alpha,beta);
disp(tau1);
tau=0:0.01:20;
sigma=sqrt(1/tau1);
mu=3;
for n=20:20:80
    x=normrnd(mu,sigma,[1,n]);
    sum=0;
    for i=1:n
        sum=sum+(x(i)-mu)^2;
    end
    pifunc=(tau./(2*pi)).^(n/2).*exp(-0.5.*sum.*tau).*gampdf(tau,alpha,beta);
    marg=[(beta^alpha)*factorial(alpha+n/2-1)]/[factorial(alpha-1)*((2*pi)^(n/2))*((beta+0.5*sum)^(n+1))];
    pifunc=pifunc/marg;
    pifunc=pifunc/max(pifunc);
    plot(tau,pifunc);
    hold on
end
hold off
xlabel('Tau=1/sigma^2');
ylabel('Posterior pdf');
legend('n=20','n=40','n=60','n=80');

%% Method of composition
mu=1;
alpha=2;
beta=3;
rng('default');
tau=gamrnd(alpha,beta);
for n=20:20:100
    sum=0;
    X=normrnd(mu,1/tau,[1,n]);
    for i=1:n
        sum=sum+(X(i)-mu)^2;
    end
    alphaNew=alpha+n/2;
    betaNew=beta+0.5*sum;
    piFunc=gampdf(0:0.5:200,alphaNew,betaNew);
    plot(0:0.5:200,piFunc);
    hold on
end
legend('n=20','n=40','n=60','n=80','n=100');
xlabel('tau');
ylabel('Posterior density');

    



    